File: qvariablestylesheet.cpp
#include "qvariablestylesheet.h"

QVariableStyleSheet::QVariableStyleSheet(QString stylesheet)
{
    _stylesheet = stylesheet;
}

void QVariableStyleSheet::setStyleSheet(QString stylesheet)
{
    _stylesheet = stylesheet;
}

QString QVariableStyleSheet::process()
{
    QString theme = _stylesheet;

    //Process CSS variables in QSS
    QRegularExpression rootExpression(":root\\s*\n*\\s*\\{\n*\\s*(--.*:.*;\n*)*\n*\\s*\\}", QRegularExpression::MultilineOption);
    QRegularExpression variableExpression("--[a-zA-Z0-9\\-\\_]+:\\s*[a-zA-Z0-9\\-\\_\\#]+;");
    QRegularExpression variableUsageExpression("var\\(--([a-zA-Z0-9\\-]*)*[\\)]{1}");
    QRegularExpression commentExpression("\\/\\*[^*]*\\*+([^/*][^*]*\\*+)*\\/");
    theme.remove(commentExpression);
    theme.replace(QRegularExpression("[\n]{2,}"), QString("\n"));
    theme.replace(QRegularExpression("[\\s]{2,}"), QString(" "));

    QRegularExpressionMatchIterator iterator = rootExpression.globalMatch(theme.toUtf8());
    QMap<QString, QString> variables;

    while(iterator.hasNext())
    {
        QRegularExpressionMatch rootMatch = iterator.next();
        QString root = rootMatch.captured(0);

        theme.remove(rootExpression);

        QRegularExpressionMatchIterator variableIterator = variableExpression.globalMatch(root.toUtf8());
        while(variableIterator.hasNext())
        {
            QRegularExpressionMatch variableMatch = variableIterator.next();
            QString variable = variableMatch.captured(0);

            QStringList split = variable.split(":");
            if(split.count() == 2)
            {
                QString key = split[0];
                QString value = split[1];

                while(value.startsWith(" "))
                    value.remove(0,1);

                while(key.startsWith(" "))
                    key.remove(0,1);
                value.remove(QRegularExpression(";$"));
                variables[key] = value;
            }
        }
    }

    iterator = variableUsageExpression.globalMatch(theme.toUtf8());

    while(iterator.hasNext())
    {
        QRegularExpressionMatch variableMatch = iterator.next();
        QString variable = variableMatch.captured(0);

        variable.remove("var(");
        variable.remove(QRegularExpression("\\)$"));

        QString replacement = variables[variable];
        theme.replace(variableMatch.captured(0), replacement);
    }

    //theme.remove(variableUsageExpression);

    return theme;
}

File: qvariablestylesheet.py
# -*- coding: utf-8 -*-
from re import compile, sub, MULTILINE


class QVariableStyleSheet(object):
    """
    QVariableStyleSheet class.
    """

    def __init__(self, stylesheet):
        """
        Initialize QVariableStyleSheet with the provided stylesheet.
        :param stylesheet: (str | unicode | QString) The stylesheet content.
        """
        self.__m_stylesheet = stylesheet  # type: str

    def setStyleSheet(self, stylesheet):
        """
        Set the stylesheet for the QVariableStyleSheet.
        :param stylesheet: (str | unicode | QString) The new stylesheet content.
        :return:
        """
        self.__m_stylesheet = stylesheet  # type: str

    def styleSheet(self):
        """
        Get the stylesheet for the QVariableStyleSheet.
        :return: str | unicode | QString
        """
        return self.__m_stylesheet

    def process(self):
        """
        Process the CSS variables in the QSS.
        :return: (str | unicode | QString) The processed stylesheet content.
        """
        theme = self.__m_stylesheet  # type: str
        # Process CSS variables in QSS.
        rootExpression = compile(r":root\s*\n*\s*\{\n*\s*(--.*:.*;\n*)*\n*\s*\}", MULTILINE)  # type: compile
        variableExpression = compile(r"--[a-zA-Z0-9\-\_]+:\s*[a-zA-Z0-9\-\_\#]+;")  # type: compile
        variableUsageExpression = compile(r"var\(--([a-zA-Z0-9\-]*)*[\)]{1}")  # type: compile
        commentExpression = compile(r"\/\*[^*]*\*+([^/*][^*]*\*+)*\/")  # type: compile
        theme = commentExpression.sub("", theme)  # type: str
        theme = sub(r"[\n]{2,}", "\n", theme)  # type: str
        theme = sub(r"[\s]{2,}", " ", theme)  # type: str
        iterator = rootExpression.finditer(theme)  # type: iter
        variables = {}  # type: dict[str, str]
        for rootMatch in iterator:
            root = rootMatch.group(0)  # type: str
            theme = rootExpression.sub("", theme)  # type: str
            variableIterator = variableExpression.finditer(root)  # type: iter
            for variableMatch in variableIterator:
                variable = variableMatch.group(0)  # type: str
                split = variable.split(":")  # type: list[str]
                if len(split) == 2:
                    key = split[0]  # type: str
                    value = split[1]  # type: str
                    while value.startswith(" "):
                        value = value[1:]  # type: str
                    while key.startswith(" "):
                        key = key[1:]  # type: str
                    value = sub(r";$", "", value)  # type: str
                    variables[key] = value  # type: str
        iterator = variableUsageExpression.finditer(theme)  # type: iter
        for variableMatch in iterator:
            variable = variableMatch.group(0)  # type: str
            variable = variable.replace("var(", "")  # type: str
            variable = sub(r"\)$", "", variable)  # type: str
            replacement = variables.get(variable, "")  # type: str
            theme = theme.replace(variableMatch.group(0), replacement)  # type: str
        # theme = sub("", theme, variableUsageExpression) # type: str
        return theme
